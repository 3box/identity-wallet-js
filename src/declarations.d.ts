// not sure why this is needed
declare module 'ipfs' {
  export default any
}

declare module 'multihashes' {
  export function encode(digest: Buffer, code: number | string, length?: number): Buffer
}
