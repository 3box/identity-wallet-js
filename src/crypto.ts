import { randomBytes } from '@stablelib/random'
export { randomBytes }
import {
  createJWE,
  decryptJWE,
  JWE,
  Encrypter,
  Decrypter,
  x25519Encrypter,
  x25519Decrypter,
} from 'did-jwt'
import { prepareCleartext, decodeCleartext } from 'dag-jose-utils'

interface EncryptOpts {
  publicKey?: Uint8Array
  encrypter?: Encrypter
  kid?: string
}

function isU8a(b: Uint8Array | undefined): b is Uint8Array {
  return b instanceof Uint8Array
}

export async function asymEncryptJWE(
  cleartext: Record<string, any>,
  { publicKey, encrypter, kid }: EncryptOpts
): Promise<JWE> {
  if (!encrypter) {
    if (!isU8a(publicKey)) throw new Error('publicKey or encrypter has to be defined')
    encrypter = x25519Encrypter(publicKey, kid)
  }
  return createJWE(prepareCleartext(cleartext), [encrypter])
}

interface DecryptOpts {
  secretKey?: Uint8Array
  decrypter?: Decrypter
}

export async function asymDecryptJWE(
  jwe: JWE,
  { secretKey, decrypter }: DecryptOpts
): Promise<Record<string, any>> {
  if (!decrypter) {
    if (!isU8a(secretKey)) throw new Error('secretKey or decrypter has to be defined')
    decrypter = x25519Decrypter(secretKey)
  }
  return decodeCleartext(await decryptJWE(jwe, decrypter))
}

interface Recipient {
  header: Record<string, string>
}

export function parseJWEKids(jwe: JWE): Array<string> {
  return (jwe.recipients || []).reduce((kids: Array<string>, recipient: Recipient) => {
    if (recipient.header?.kid) kids.push(recipient.header.kid.split('#')[1])
    return kids
  }, [])
}
