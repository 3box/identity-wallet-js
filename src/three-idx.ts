import { CeramicApi, Doctype, CeramicRecord } from '@ceramicnetwork/common'
import { TileDoctype } from '@ceramicnetwork/doctype-tile'
import CeramicClient from '@ceramicnetwork/http-client'
import { schemas, definitions } from '@ceramicstudio/idx-constants'
import CID from 'cids'

import type { DidProvider } from './did-provider'
import type { ThreeIdState } from './keyring'
import type { DID } from 'dids'
import type { JWE } from 'did-jwt'
import type DocID from '@ceramicnetwork/docid'

const KEYCHAIN_DEF = definitions.threeIdKeychain
const IDX = 'IDX'
const { IdentityIndex, ThreeIdKeychain } = schemas

const isLegacyDid = (didId: string): boolean => {
  try {
    new CID(didId)
    return true
  } catch (e) {
    return false
  }
}

export interface EncData {
  jwe?: JWE
}

export interface EncKeyMaterial {
  seed: EncData
  pastSeeds: Array<JWE>
}

export interface AuthEntry {
  data: EncData
  id: EncData
  //linkDoc: string
  //pub: string // base58 multicodec x25519 public key
}

interface AuthMap {
  [did: string]: AuthEntry
}

export interface NewAuthEntry {
  mapEntry: AuthMap
  did: DID
}

interface AuthLinkDocUpdate {
  record: CeramicRecord
  docid: DocID
  did: string
}

export class ThreeIDX {
  public docs: Record<string, Doctype>
  public ceramic: CeramicApi
  protected _v03ID?: string

  constructor(ceramic?: CeramicApi) {
    this.ceramic = ceramic || new CeramicClient()
    this.docs = {}
  }

  async setDIDProvider(provider: DidProvider): Promise<void> {
    await this.ceramic.setDIDProvider(provider)
  }

  setV03ID(did: string): void {
    this._v03ID = did
  }

  get id(): string {
    return this._v03ID || `did:3:${this.docs.threeId.id.baseID.toString()}`
  }

  async create3idDoc(docParams: Record<string, any>): Promise<void> {
    this.docs.threeId = await this.ceramic.createDocument('tile', docParams, {
      anchor: false,
      publish: false,
    })
  }

  get3idVersion(): string {
    const docId = this.docs.threeId.anchorCommitIds.pop()
    return docId ? (docId.commit as CID).toString() : '0'
  }

  async loadDoc(name: string, controller: string, family: string): Promise<void> {
    this.docs[name] = await this.ceramic.createDocument(
      'tile',
      { deterministic: true, metadata: { controllers: [controller], family } },
      { anchor: false, publish: false }
    )
  }

  async createAuthLinkUpdate({ did }: NewAuthEntry): Promise<AuthLinkDocUpdate> {
    const didString = did.id
    await this.loadDoc(didString, didString, 'authLink')
    await this.ceramic.pin.add(this.docs[didString].id)
    return {
      record: await TileDoctype._makeRecord(this.docs[didString], did, { did: this.id }),
      docid: this.docs[didString].id,
      did: didString,
    }
  }

  async applyAuthLinkUpdate({ docid, record, did }: AuthLinkDocUpdate): Promise<void> {
    if (this.docs[did].content !== this.id) {
      await this.ceramic.applyRecord(docid, record)
    }
  }

  /**
   * Create a new IDX structure that has a given authEntry in it's keychain.
   */
  async createIDX(newEntry?: NewAuthEntry): Promise<void> {
    const docUpdatePromise = newEntry ? this.createAuthLinkUpdate(newEntry) : Promise.resolve(null)
    // eslint-disable-next-line prettier/prettier
    await Promise.all([
      this.loadDoc(KEYCHAIN_DEF, this.id, KEYCHAIN_DEF),
      this.loadDoc('idx', this.id, IDX),
    ])
    // eslint-disable-next-line prettier/prettier
    await Promise.all([
      this.pinAllDocs(),
      this.updateKeychainDoc(newEntry?.mapEntry),
      this.addKeychainToIDX(),
    ])
    // Only update the link document after the keychain have been updated.
    const docUpdate = await docUpdatePromise
    if (docUpdate) {
      await this.applyAuthLinkUpdate(docUpdate)
    }
  }

  /**
   * Returns the encrypted JWE for the given authLink
   */
  async loadIDX(authDid: string): Promise<EncKeyMaterial | null> {
    await this.loadDoc(authDid, authDid, 'authLink')
    const { did } = this.docs[authDid].content
    if (!did) return null
    // eslint-disable-next-line prettier/prettier
    await Promise.all([
      this.load3IDDoc(did),
      this.loadDoc(KEYCHAIN_DEF, did, KEYCHAIN_DEF),
      this.loadDoc('idx', did, IDX),
    ])
    const { authMap, pastSeeds } = this.docs[KEYCHAIN_DEF].content
    return {
      seed: authMap[authDid]?.data,
      pastSeeds,
    } as EncKeyMaterial
  }

  async load3IDDoc(did: string): Promise<void> {
    const id = did.split(':')[2]
    if (isLegacyDid(id)) {
      // we have to load the document later when keys are loaded
      this._v03ID = did
    } else {
      this.docs.threeId = await this.ceramic.loadDocument(id)
    }
  }

  /**
   * Reset the IDX doc structure to a default (mostly empty) state.
   */
  async resetIDX(): Promise<void> {
    if (this.docs.idx == null) {
      throw new Error('No IDX doc')
    }
    const update: Record<string, any> = {
      content: { [KEYCHAIN_DEF]: this.docs[KEYCHAIN_DEF].id.toUrl() },
    }
    if (!this.docs.idx.metadata.schema) {
      update.metadata = { schema: IdentityIndex }
    }
    await this.docs.idx.change(update)
  }

  /**
   * Adds a new AuthEntries to the Auth keychain.
   */
  async addAuthEntries(newEntries: Array<NewAuthEntry>): Promise<void> {
    const linkDocUpdatesPromise = Promise.all(newEntries.map(this.createAuthLinkUpdate.bind(this)))
    const { authMap, pastSeeds } = this.docs[KEYCHAIN_DEF].content
    const newAuthEntries = newEntries.reduce(
      (acc, { mapEntry }) => Object.assign(acc, mapEntry),
      {}
    )
    Object.assign(authMap, newAuthEntries)
    await this.updateKeychainDoc(authMap, pastSeeds)
    await Promise.all((await linkDocUpdatesPromise).map(this.applyAuthLinkUpdate.bind(this)))
  }

  /**
   * Returns all public keys that is in the auth keychain.
   */
  getAuthMap(): AuthMap {
    if (!this.docs[KEYCHAIN_DEF] || !this.docs[KEYCHAIN_DEF].content.authMap) return {}
    return this.docs[KEYCHAIN_DEF].content.authMap as AuthMap
  }

  async pinAllDocs(): Promise<void> {
    await Promise.all(
      Object.values(this.docs).map(async (doc) => {
        await this.ceramic.pin.add(doc.id)
      })
    )
  }

  async addKeychainToIDX(): Promise<void> {
    const content = this.docs.idx.content
    if (!content || !content[KEYCHAIN_DEF]) {
      const update: Record<string, any> = {
        content: Object.assign(content || {}, {
          [KEYCHAIN_DEF]: this.docs[KEYCHAIN_DEF].id.toUrl(),
        }),
      }
      if (!this.docs.idx.metadata.schema) {
        update.metadata = { schema: IdentityIndex }
      }
      await this.docs.idx.change(update)
    }
  }

  async updateKeychainDoc(authMap: AuthMap = {}, pastSeeds: Array<JWE> = []): Promise<void> {
    if (Object.keys(authMap).length !== 0) {
      const update: Record<string, any> = { content: { authMap, pastSeeds } }
      if (!this.docs[KEYCHAIN_DEF].metadata.schema) {
        update.metadata = { schema: ThreeIdKeychain }
      }
      await this.docs[KEYCHAIN_DEF].change(update)
    }
  }

  /**
   * Preform a key rotation.
   * Will update the keys in the 3id document, and create a new 3ID keychain
   * with the given authEntries.
   */
  async rotateKeys(
    threeIdState: ThreeIdState,
    pastSeeds: Array<JWE>,
    authMap: AuthMap
  ): Promise<void> {
    if (!threeIdState.content) throw new Error('Content has to be defined')
    // Rotate keys in 3ID document and update keychain
    await Promise.all([
      this.docs.threeId.change({
        metadata: threeIdState.metadata,
        content: { ...this.docs.threeId.content, publicKeys: threeIdState.content.publicKeys },
      }),
      this.updateKeychainDoc(authMap, pastSeeds),
      this.pinAllDocs(),
    ])
  }
}
