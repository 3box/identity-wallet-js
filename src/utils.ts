import { sha256 } from 'js-sha256'
import Multihash from 'multihashes'
import bs58 from 'bs58'
import { Wallet } from '@ethersproject/wallet'
import stringify from 'fast-json-stable-stringify'

const ENC_BLOCK_SIZE = 24

export interface PublicKeys {
  signingKey: string
  managementKey: string | null
  asymEncryptionKey: string
}

export const pad = (val: string, blockSize = ENC_BLOCK_SIZE): string => {
  const blockDiff = (blockSize - (val.length % blockSize)) % blockSize
  return `${val}${'\0'.repeat(blockDiff)}`
}

export const unpad = (padded: string): string => padded.replace(/\0+$/, '')

export const sha256Multihash = (str: string): string => {
  return Multihash.encode(Buffer.from(sha256(str)), 'sha2-256').toString('hex')
}

const multicodecPubkeyTable: Record<string, number> = {
  secp256k1: 0xe7,
  x25519: 0xec,
  ed25519: 0xed,
}

export function encodeKey(key: Uint8Array, keyType: string): string {
  const buf = new Uint8Array(key.length + 2)
  if (!multicodecPubkeyTable[keyType]) {
    throw new Error(`Key type "${keyType}" not supported.`)
  }
  buf[0] = multicodecPubkeyTable[keyType]
  // The multicodec is encoded as a varint so we need to add this.
  // See js-multicodec for a general implementation
  buf[1] = 0x01
  buf.set(key, 2)
  return `z${bs58.encode(buf)}`
}

export const fakeEthProvider = (wallet: Wallet) => ({
  send: (
    request: { method: string; params: Array<any> },
    callback: (err: Error | null | undefined, res?: any) => void
  ) => {
    if (request.method !== 'personal_sign') {
      callback(new Error('only supports personal_sign'))
    } else {
      let message = request.params[0]
      if (message.startsWith('0x')) {
        message = Buffer.from(message.slice(2), 'hex').toString('utf8')
      }
      callback(null, { result: wallet.signMessage(message) })
    }
  },
})

export function hexToU8A(str: string): Uint8Array {
  return new Uint8Array(Buffer.from(str, 'hex'))
}

export function toStableObject(obj: any): any {
  return JSON.parse(stringify(obj))
}
