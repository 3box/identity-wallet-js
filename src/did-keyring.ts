import nacl, { BoxKeyPair } from 'tweetnacl'
import naclutil from 'tweetnacl-util'
import { HDNode } from '@ethersproject/hdnode'
import { Wallet } from '@ethersproject/wallet'
import { SimpleSigner } from 'did-jwt'
import { sha256 } from 'js-sha256'
import { ec as EC } from 'elliptic'

import { naclRandom } from './crypto'

const ec = new EC('secp256k1')

const BASE_PATH = "m/51073068'/0'"
const ROOT_STORE_PATH = "0'/0'/0'/0'/0'/0'/0'/0'"

const AUTH_PATH_WALLET = BASE_PATH + '/' + ROOT_STORE_PATH + '/0'
const AUTH_PATH_ENCRYPTION = BASE_PATH + '/' + ROOT_STORE_PATH + '/3'

const ensure0x = (str: string): string => {
  return (str.startsWith('0x') ? '' : '0x') + str
}

interface KeysSet {
  signingKey: HDNode
  asymEncryptionKey: BoxKeyPair
  symEncryptionKey: Uint8Array
  managementKey: HDNode
}

export class Keyring {
  public static create() {
    const seed = '0x' + Buffer.from(naclRandom(32)).toString('hex')
    return new Keyring(seed)
  }

  private _seed: string
  private _baseNode: HDNode
  private _keys: KeysSet

  constructor(seed: string) {
    this._seed = seed
    this._baseNode = HDNode.fromSeed(seed).derivePath(BASE_PATH)
    const rootNode = this._baseNode.derivePath(ROOT_STORE_PATH)
    this._keys = {
      signingKey: rootNode.derivePath('0'),
      asymEncryptionKey: nacl.box.keyPair.fromSecretKey(
        new Uint8Array(
          Buffer.from(rootNode.derivePath('2').privateKey.slice(2), 'hex'),
        ),
      ),
      symEncryptionKey: Keyring.hexToUint8Array(
        rootNode.derivePath('3').privateKey.slice(2),
      ),
      managementKey: rootNode.derivePath('1'),
    }
  }

  _getKeys() {
    return this._keys
  }

  asymEncrypt(msg, toPublic, { nonce } = {}) {
    nonce = nonce || Keyring.randomNonce()
    toPublic = naclutil.decodeBase64(toPublic)
    if (typeof msg === 'string') {
      msg = naclutil.decodeUTF8(msg)
    }
    const ephemneralKeypair = nacl.box.keyPair()
    const ciphertext = nacl.box(
      msg,
      nonce,
      toPublic,
      ephemneralKeypair.secretKey,
    )
    return {
      nonce: naclutil.encodeBase64(nonce),
      ephemeralFrom: naclutil.encodeBase64(ephemneralKeypair.publicKey),
      ciphertext: naclutil.encodeBase64(ciphertext),
    }
  }

  asymDecrypt(ciphertext, fromPublic, nonce, { space, toBuffer } = {}) {
    fromPublic = naclutil.decodeBase64(fromPublic)
    ciphertext = naclutil.decodeBase64(ciphertext)
    nonce = naclutil.decodeBase64(nonce)
    const cleartext = nacl.box.open(
      ciphertext,
      nonce,
      fromPublic,
      this._keys.asymEncryptionKey.secretKey,
    )
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  symEncrypt(msg, { space, nonce } = {}) {
    return Keyring.symEncryptBase(msg, this._keys.symEncryptionKey, nonce)
  }

  symDecrypt(ciphertext, nonce, { space, toBuffer } = {}) {
    return Keyring.symDecryptBase(
      ciphertext,
      this._keys.symEncryptionKey,
      nonce,
      toBuffer,
    )
  }

  managementPersonalSign(message) {
    const wallet = this.managementWallet()
    return wallet.signMessage(message)
  }

  managementWallet() {
    return new Wallet(this._keys.managementKey.privateKey)
  }

  getJWTSigner() {
    return SimpleSigner(this._keys.signingKey.privateKey.slice(2))
  }

  getDBSalt() {
    return sha256(this._keys.signingKey.derivePath('0').privateKey.slice(2))
  }

  getPublicKeys({ uncompressed }) {
    let signingKey = this._keys.signingKey.publicKey.slice(2)
    const managementKey = this._keys.managementKey.address
    if (uncompressed) {
      signingKey = ec
        .keyFromPublic(Buffer.from(signingKey, 'hex'))
        .getPublic(false, 'hex')
    }
    return {
      signingKey,
      managementKey,
      asymEncryptionKey: naclutil.encodeBase64(
        this._keys.asymEncryptionKey.publicKey,
      ),
    }
  }

  serialize() {
    return this._seed
  }

  static encryptWithAuthSecret(message, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symEncryptBase(message, key)
  }

  static decryptWithAuthSecret(ciphertext, nonce, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symDecryptBase(ciphertext, key, nonce)
  }

  static walletForAuthSecret(authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_WALLET,
    )
    return new Wallet(node.privateKey)
  }

  static hexToUint8Array(str) {
    return new Uint8Array(Buffer.from(str, 'hex'))
  }

  static symEncryptBase(msg, symKey, nonce) {
    nonce = nonce || Keyring.randomNonce()
    if (typeof msg === 'string') {
      msg = nacl.util.decodeUTF8(msg)
    }
    const ciphertext = nacl.secretbox(msg, nonce, symKey)
    return {
      nonce: nacl.util.encodeBase64(nonce),
      ciphertext: nacl.util.encodeBase64(ciphertext),
    }
  }

  static symDecryptBase(ciphertext, symKey, nonce, toBuffer) {
    ciphertext = nacl.util.decodeBase64(ciphertext)
    nonce = nacl.util.decodeBase64(nonce)
    const cleartext = nacl.secretbox.open(ciphertext, nonce, symKey)
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  static naclRandom(length) {
    return nacl.randomBytes(length)
  }

  static randomNonce() {
    return Keyring.naclRandom(24)
  }
}

module.exports = Keyring
