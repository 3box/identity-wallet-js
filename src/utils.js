import crypto from 'crypto'
import Multihash from 'multihashes'

const ENC_BLOCK_SIZE = 24

const pad = (val, blockSize = ENC_BLOCK_SIZE) => {
  const blockDiff = (blockSize - (val.length % blockSize)) % blockSize
  return `${val}${'\0'.repeat(blockDiff)}`
}

const unpad = padded => padded.replace(/\0+$/, '')

const sha256 = msg => crypto.createHash('sha256').update(msg).digest('hex')

const sha256Multihash = str => Multihash.encode(Buffer.from(sha256(str)), 'sha2-256').toString('hex')

module.exports = {
  pad,
  unpad,
  sha256,
  sha256Multihash
}
