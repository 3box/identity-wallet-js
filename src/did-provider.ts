import { createJWS } from 'did-jwt'
import {
  HandlerMethods,
  RequestHandler,
  RPCConnection,
  RPCError,
  RPCRequest,
  RPCResponse,
  createHandler,
} from 'rpc-utils'

import Keyring from './keyring'
import { ThreeIDX } from './three-idx'
import Permissions from './permissions'
import { toStableObject } from './utils'

type Origin = string | null | undefined

export type Context = {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  origin: Origin
}

interface CreateJWSParams {
  payload: Record<string, any>
  protected?: Record<string, any>
  did: string
}

interface AuthParams {
  paths: Array<string>
}

export const didMethods: HandlerMethods<Context> = {
  did_authenticate: async ({ permissions, threeIdx, origin }, params: AuthParams) => {
    const paths = await permissions.request(origin, params.paths || [])
    // paths should be an array if permission granted
    // may be a subset or requested paths or empty array
    if (paths === null) throw new RPCError(4001, 'User Rejected Request')
    return { did: threeIdx.DID, paths }
  },
  did_createJWS: async ({ permissions, keyring, threeIdx, origin }, params: CreateJWSParams) => {
    if (!permissions.has(origin)) {
      throw new RPCError(4100, 'Unauthorized')
    }
    // TODO - if the requesting DID is our management key
    // (did:key) we should request explicit permission.
    const keyName = threeIdx.parseKeyName(params.did)
    const kid = await threeIdx.encodeKidWithVersion(keyName)
    const signer = keyring.getSigner(keyName)
    const header = toStableObject(Object.assign(params.protected || {}, { kid }))
    const jws = await createJWS(toStableObject(params.payload), signer, header)
    return { jws }
  },
}

export interface ProviderConfig {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  forcedOrigin?: string
}

export class DidProvider implements RPCConnection {
  protected _handle: RequestHandler

  constructor({ permissions, threeIdx, keyring, forcedOrigin }: ProviderConfig) {
    const handler = createHandler<Context>(didMethods)
    this._handle = (origin: string, msg: RPCRequest) => {
      return handler(
        {
          origin: forcedOrigin || origin,
          permissions,
          threeIdx,
          keyring,
        },
        msg
      )
    }
  }

  public get isDidProvider(): boolean {
    return true
  }

  public async send(msg: RPCRequest, origin?: Origin): Promise<RPCResponse | null> {
    return await this._handle(origin, msg)
  }
}
