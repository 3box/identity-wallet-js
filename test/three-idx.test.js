import tmp from 'tmp-promise'
import Ceramic from '@ceramicnetwork/ceramic-core'
import Ipfs from 'ipfs'
import all from 'it-all'
import { AccountID } from 'caip'
import { createLink } from '3id-blockchain-utils'

import { ThreeIDX } from '../src/three-idx'
import { DidProvider } from '../src/did-provider'
import Keyring from '../src/keyring'
import {
  AsymEncryptedMessage,
  asymDecrypt,
  asymEncrypt,
  naclRandom,
} from '../src/crypto'
import { fakeEthProvider } from '../src/utils'

import dagJose from 'dag-jose'
import basicsImport from 'multiformats/cjs/src/basics-import.js'
import legacy from 'multiformats/cjs/src/legacy.js'

const seed = '0x8e641c0dc77f6916cc7f743dad774cdf9f6f7bcb880b11395149dd878377cd398650bbfd4607962b49953c87da4d7f3ff247ed734b06f96bdd69479377bc612b'

const genIpfsConf = (folder) => {
  basicsImport.multicodec.add(dagJose)
  const format = legacy(basicsImport, dagJose.name)
  return {
    ipld: { formats: [format] },
    repo: `${folder}/ipfs/`,
    config: {
      Addresses: { Swarm: [] },
      Bootstrap: []
    },
  }
}

const randomSecret = () => '0x' + Buffer.from(naclRandom(32)).toString('hex')

const genAuthEntryCreate = async (did) => {
  const wallet = Keyring.authSecretToWallet(randomSecret())
  const accountId = new AccountID({ address: wallet.address, chainId: 'eip155:1' })
  const newAuthEntry = {
    pub: 'publickey' + randomSecret(),
    data: 'authdata' + randomSecret(),
    id: 'authid' + randomSecret(),
    linkProof: await createLink(did || 'did:3:asdf', accountId, fakeEthProvider(wallet))
  }
  return { newAuthEntry, accountId: accountId.toString() }
}

const setup3id = async (threeIdx, keyring) => {
  const pubkeys = keyring.getPublicKeys({ mgmtPub: true, useMulticodec: true })
  const forcedDID = `did:key:${pubkeys.managementKey}`
  let didProvider = new DidProvider({ permissions: mockedPermissions, threeIdx, keyring, forcedDID })
  await threeIdx.setDIDProvider(didProvider)
  await threeIdx.create3idDoc(pubkeys)
  didProvider = new DidProvider({ permissions: mockedPermissions, threeIdx, keyring })
  await threeIdx.setDIDProvider(didProvider)
}

const mockedPermissions = {
  request: async () => [],
  has: () => true,
}

describe('ThreeIDX', () => {
  jest.setTimeout(15000)
  let tmpFolder
  let ipfs, ceramic
  let keyring, threeIdx

  beforeAll(async () => {
    tmpFolder = await tmp.dir({ unsafeCleanup: true })
    ipfs = await Ipfs.create(genIpfsConf(tmpFolder.path))
    ceramic = await Ceramic.create(ipfs, { stateStorePath: tmpFolder.path + '/ceramic/'})
  })

  afterAll(async () => {
    await ceramic.close()
    await ipfs.stop()
    await tmpFolder.cleanup()
  })

  beforeEach(async () => {
    keyring = new Keyring(randomSecret())
    threeIdx = new ThreeIDX(ceramic)
  })

  it('creates 3id doc', async () => {
    keyring = new Keyring(seed)
    await setup3id(threeIdx, keyring)
    expect(threeIdx.docs.threeId.state).toMatchSnapshot()
  })

  it('parses key name correctly', async () => {
    await setup3id(threeIdx, keyring)
    const badKid = 'did:3:bayfiuherg98h349h#signing'
    expect(() => threeIdx.parseKeyName(badKid)).toThrow('Invalid DID')
    const goodKid = threeIdx.id + '#signing'
    expect(threeIdx.parseKeyName(goodKid)).toEqual('signing')
  })

  it('encodes kid with version', async () => {
    await setup3id(threeIdx, keyring)
    // with no anchor
    expect(await threeIdx.encodeKidWithVersion()).toEqual(threeIdx.id + '?version-id=0#signing')
    expect(await threeIdx.encodeKidWithVersion('management')).toEqual(`${threeIdx.managementDID}#${threeIdx.managementDID.split(':')[2]}`)
    // wait for anchor
    await new Promise(resolve => threeIdx.docs.threeId.on('change', resolve))
    const latestVer = (await ceramic.listVersions(threeIdx.docs.threeId.id)).pop()
    expect(await threeIdx.encodeKidWithVersion()).toEqual(threeIdx.id + '?version-id=' + latestVer + '#signing')
  })

  it('creates authMapEntry', async () => {
    const { newAuthEntry, accountId } = await genAuthEntryCreate()
    const authMapEntry = await threeIdx.createAuthMapEntry(newAuthEntry)

    expect(authMapEntry).toEqual({
      [threeIdx.docs[accountId].id]: {
        pub: newAuthEntry.pub,
        data: newAuthEntry.data,
        id: newAuthEntry.id,
      }
    })
    expect(threeIdx.docs[accountId].owners).toEqual([accountId])
    expect(threeIdx.docs[accountId].content).toEqual('did:3:asdf')
  })

  it('createIDX with new auth entry', async () => {
    await setup3id(threeIdx, keyring)
    const { newAuthEntry, accountId } = await genAuthEntryCreate()
    await threeIdx.createIDX(newAuthEntry)

    expect(threeIdx.docs['auth-keychain'].content).toEqual({
      [threeIdx.docs[accountId].id]: {
        pub: newAuthEntry.pub,
        data: newAuthEntry.data,
        id: newAuthEntry.id,
      }
    })
    expect(threeIdx.docs.idx.content).toEqual({ 'auth-keychain': threeIdx.docs['auth-keychain'].id })
    expect(threeIdx.docs.threeId.content).toEqual(expect.objectContaining({ 'idx': threeIdx.docs.idx.id }))
    // should be pinned
    expect(await all(await ceramic.pin.ls())).toEqual(expect.arrayContaining([
      threeIdx.docs.threeId.id,
      threeIdx.docs.idx.id,
      threeIdx.docs['auth-keychain'].id,
      threeIdx.docs[accountId].id,
    ].map(docid => docid.replace('ceramic://', '/ceramic/'))))
  })

  it('createIDX with no auth entry', async () => {
    await setup3id(threeIdx, keyring)
    await threeIdx.createIDX()

    expect(threeIdx.docs.idx.content).toEqual({ 'auth-keychain': threeIdx.docs['auth-keychain'].id })
    expect(threeIdx.docs.threeId.content).toEqual(expect.objectContaining({ 'idx': threeIdx.docs.idx.id }))
    // should be pinned
    expect(await all(await ceramic.pin.ls())).toEqual(expect.arrayContaining([
      threeIdx.docs.threeId.id,
      threeIdx.docs.idx.id,
    ].map(docid => docid.replace('ceramic://', '/ceramic/'))))
  })

  it('loadIDX fails if authLink does not exist', async () => {
    await setup3id(threeIdx, keyring)
    const { newAuthEntry, accountId } = await genAuthEntryCreate(threeIdx.id)

    expect(await threeIdx.loadIDX(accountId)).toEqual(null)
  })

  it('loadIDX works if IDX created', async () => {
    await setup3id(threeIdx, keyring)
    const { newAuthEntry, accountId } = await genAuthEntryCreate(threeIdx.id)
    await threeIdx.createIDX(newAuthEntry)

    expect(await threeIdx.loadIDX(accountId)).toEqual(newAuthEntry.data)
  })

  it('addAuthEntries', async () => {
    await setup3id(threeIdx, keyring)
    const { newAuthEntry: nae1, accountId: ai1 } = await genAuthEntryCreate(threeIdx.id)
    await threeIdx.createIDX(nae1)

    const authEntry1 = { pub: nae1.pub, data: nae1.data, id: nae1.id }
    expect(threeIdx.getAllAuthEntries()).toEqual([authEntry1])

    const { newAuthEntry: nae2, accountId: ai2 } = await genAuthEntryCreate(threeIdx.id)
    const { newAuthEntry: nae3, accountId: ai3 } = await genAuthEntryCreate(threeIdx.id)
    const authEntry2 = { pub: nae2.pub, data: nae2.data, id: nae2.id }
    const authEntry3 = { pub: nae3.pub, data: nae3.data, id: nae3.id }
    await threeIdx.addAuthEntries([nae2, nae3])

    expect(threeIdx.getAllAuthEntries()).toEqual([authEntry1, authEntry2, authEntry3])
    expect(await all(await ceramic.pin.ls())).toEqual(expect.arrayContaining([
      threeIdx.docs[ai1].id,
      threeIdx.docs[ai2].id,
      threeIdx.docs[ai3].id,
    ].map(docid => docid.replace('ceramic://', '/ceramic/'))))
  })
})
