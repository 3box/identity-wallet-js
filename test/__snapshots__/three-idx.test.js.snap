// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeIDX creates 3id doc 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "publicKeys": Object {
      "5rFyC95qwGXUokA": "zQ3shTGj6xVzEsyR5xRr1JQ6nm3CHBrws85rFyC95qwGXUokA",
      "YgmQJCC4FPsiSVf": "z6LSfM86nM4fYaHyUbRVDEG1itNw4j9GHYgmQJCC4FPsiSVf",
    },
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "codec": "dag-jose",
      "hash": Uint8Array [
        18,
        32,
        224,
        91,
        122,
        238,
        150,
        222,
        162,
        223,
        157,
        200,
        205,
        74,
        194,
        13,
        121,
        201,
        146,
        228,
        71,
        140,
        41,
        33,
        65,
        218,
        35,
        95,
        8,
        123,
        104,
        207,
        23,
        172,
      ],
      "version": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:key:zQ3shggrPaEwrC8tcvUwg1bwUPAos4XLXPUUH5uFcCnjSkqAm",
    ],
    "family": "3id",
  },
  "signature": 2,
}
`;
